src/blockchain/fork_choice.cpp

```diff
@@ -482,6 +483,64 @@ namespace lean {
     return true;
   }
 
+  void ForkChoiceStore::updateLastFinalized(const Checkpoint& checkpoint) {
+    BOOST_ASSERT(checkpoint.slot > latest_finalized_.slot);
+    latest_finalized_ = checkpoint;
+
+    pruneStatesAfterFinalized();
+
+    // Safety: pull up head/safe_target up to the finalized root (just in case).
+    if (not blocks_.contains(head_)) {
+      head_ = latest_finalized_.root;
+    }
+    if (not blocks_.contains(safe_target_)) {
+      safe_target_ = latest_finalized_.root;
+    }
+  }
+
+  void ForkChoiceStore::pruneStatesAfterFinalized() {
+    const auto& finalized_root = latest_finalized_.root;
+    const auto& finalized_slot = latest_finalized_.slot;
+
+    // Collect block hashes to erase
+    std::vector<BlockHash> to_erase;
+    to_erase.reserve(blocks_.size());
+
+    for (const auto &[hash, signed_block] : blocks_) {
+      const auto &block = signed_block.message.block;
+
+      if (hash == finalized_root) {
+        continue;
+      }
+      if (block.slot < finalized_slot) {
+        to_erase.push_back(hash);
```

Comment:
Do you want to remove **all** blocks older than finalized block?
Did you want to remove forks (traverse, maybe cache map from parent hash to children list)?